lib/pacemaker/pcmk_sched_group.c

```diff
@@ -61,20 +64,23 @@ expand_group_colocations(pe_resource_t *rsc)
                 pcmk__colocation_t *constraint = (pcmk__colocation_t *) cons_iter->data;
 
                 if (constraint->score == INFINITY) {
-                    member->rsc_cons = g_list_prepend(member->rsc_cons, constraint);
+                    pcmk__add_this_with(member, constraint);
                 }
             }
         } else if (!pcmk_is_set(member->flags, pe_rsc_managed)) {
             any_unmanaged = true;
         }
     }
 
+    g_list_free(rsc->rsc_cons);
```

Comment:
This only frees the list and doesn't touch the elements, right? AFAICT only `g_list_free_full()` frees the elements, but the doc says "The freed elements are returned to the slice allocator." Not sure off hand what that means.

https://docs.gtk.org/glib/type_func.List.free.html

Reply:
Correct, it's a little obscure here: pe_working_set_t has a complete list of all colocation constraints in the cluster, then each resource has the constraints affecting it added to its rsc_cons or rsc_cons_lhs. Therefore, we free only the lists, not the constraints themselves, when freeing rsc_cons and rsc_cons_lhs; the constraints are freed when the pe_working_set_t is freed.